lang/java/reef-common/src/main/java/org/apache/reef/util/JARFileMaker.java

```diff
@@ -34,87 +34,83 @@
 
   private static final Logger LOG = Logger.getLogger(JARFileMaker.class.getName());
 
-  private final FileOutputStream fileOutputStream;
   private final JarOutputStream jarOutputStream;
-  private String relativeStartCanonicalPath = null;
 
-  public JARFileMaker(final File outputFile, final Manifest manifest) throws IOException {
-    this.fileOutputStream = new FileOutputStream(outputFile);
-    this.jarOutputStream = new JarOutputStream(this.fileOutputStream, manifest);
+  public JARFileMaker(final File outputFile) throws IOException {
+    this(outputFile, null);
   }
 
-  public JARFileMaker(final File outputFile) throws IOException {
-    this.fileOutputStream = new FileOutputStream(outputFile);
-    this.jarOutputStream = new JarOutputStream(this.fileOutputStream);
+  public JARFileMaker(final File outputFile, final Manifest manifest) throws IOException {
+    LOG.log(Level.FINER, "Output jar: {0}", outputFile);
+    final FileOutputStream outputStream = new FileOutputStream(outputFile);
+    this.jarOutputStream = manifest == null ?
+        new JarOutputStream(outputStream) : new JarOutputStream(outputStream, manifest);
   }
 
   /**
    * Adds a file to the JAR. If inputFile is a folder, it will be added recursively.
-   *
-   * @param inputFile
-   * @throws IOException
+   * @param inputFile file or directory to be added to the jar.
+   * @throws IOException if cannot create a jar.
    */
   public JARFileMaker add(final File inputFile) throws IOException {
+    return this.add(inputFile, null);
+  }
 
-    final String fileNameInJAR = makeRelative(inputFile);
-    if (inputFile.isDirectory()) {
-      final JarEntry entry = new JarEntry(fileNameInJAR);
-      entry.setTime(inputFile.lastModified());
-      this.jarOutputStream.putNextEntry(entry);
-      this.jarOutputStream.closeEntry();
-      final File[] files = inputFile.listFiles();
-      if (files != null) {
-        for (final File nestedFile : files) {
-          add(nestedFile);
-        }
+  public JARFileMaker addChildren(final File folder) throws IOException {
+    LOG.log(Level.FINEST, "Add children: {0}", folder);
+    final File[] files = folder.listFiles();
+    if (files != null) {
+      for (final File nestedFile : files) {
```

Comment:
We could make this slightly nicer using `commons-lang` `ArrayUtils`:

```java
for (final File nestedFile :  ArrayUtils.nullToEmpty(files)) {
...
}
```

This way, we don't need the `if` anymore.